parent_visitor.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2017 Ernest Micklei
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package proto
+
+import "log"
+
+// ParentAwareVisitor keeps track of all the parents of each AST node
+// while visiting all the elements and options.
+type ParentAwareVisitor struct {
+	Parents  []Visitee
+	delegate Visitor
+}
+
+// NewParentAwareVisitor returns a ParentAwareVisitor that collects the parents
+// of each visited and delegates the Visit call to the delegate.
+func NewParentAwareVisitor(delegate Visitor) *ParentAwareVisitor {
+	return &ParentAwareVisitor{delegate: delegate}
+}
+
+func (v *ParentAwareVisitor) push(parent Visitee) {
+	v.Parents = append(v.Parents, parent)
+}
+
+func (v *ParentAwareVisitor) pop() {
+	if len(v.Parents) == 0 {
+		log.Fatal("no parent to pop")
```

Comment:
This isn't very lib-like - most people don't use the log package, and this is an uncatchable `os.Exit` call.

This single line would prevent me from using all this code.

Reply:
you are right. actually this was debug code added to get a better hint at what went wrong when building it.  I think it can be removed altogether.